src/core/net/ip6.cpp

```diff
@@ -1268,6 +1285,22 @@ otError Ip6::HandleDatagram(Message &aMessage, Netif *aNetif, const void *aLinkM
             hopLimit = header.GetHopLimit();
             aMessage.Write(Header::GetHopLimitOffset(), Header::GetHopLimitSize(), &hopLimit);
 
+            // check whether source port is an unsecure port
+            if (aFromNcpHost && (nextHeader == kProtoTcp || nextHeader == kProtoUdp))
+            {
+                uint16_t sourcePort;
+
+                VerifyOrExit(aMessage.Read(aMessage.GetOffset(), sizeof(sourcePort), &sourcePort) == sizeof(sourcePort),
+                             error = OT_ERROR_PARSE);
+                sourcePort = HostSwap16(sourcePort);
+                if (Get<ot::Ip6::Filter>().IsPortUnsecure(sourcePort))
+                {
+                    aMessage.SetLinkSecurityEnabled(false);
+                    otLogInfoIp6("Disabled link security for packet to %s",
+                                 header.GetDestination().ToString().AsCString());
```

Comment:
This is rather big change in how the unsecure port is used compared to before.

Unseucre port was only related to **allowing rx** of IPv6 message (link-local address) from the port number (without link layer security). This change now applies the port to the **IPv6 message transmissions** (to same port) as well.

I still have the same concern on this that this code now overrides (removes) the link security on the message even if the user has explicitly requested it (if the port matches the unsecure port).  I personally feel we should follow what host/app layer and not override/reduce the security-level (keep the logic as before).

I this it may be ok  if we do it the other way (add link-layer security when user did not request it), i.e. check if  user has asked the message to be sent without link-security but the destination is an attached neighbor, then require security.  

Thoughts?

Reply:
One idea here is to add a new flag/option in `otMessageSettings` related to link-layer security. Basically have 3 options:
- Host/user requires link-layer security
- Host/user explicitly asks for no link-layer security
- Host/user delegates it to lower level stack

We can then apply the rule about the unsecure port only if user delegated it to stack. 

Thoughts?